services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionAsyncClient.java

```diff
@@ -15,80 +15,131 @@
+
+        Optional<String> bucketRegionFromException = getBucketRegionFromException((S3Exception) t.getCause());
+        if (bucketRegionFromException.isPresent()) {
+            sendRequestWithRightRegion(request, operation, bucketName, returnFuture,
+                                       bucketRegionFromException);
+        } else {
+            fetchRegionAndSendRequest(request, operation, bucketName, returnFuture);
+        }
     }
 
-    //TODO: optimize shared sync/async code
-    private AwsRequestOverrideConfiguration getOrCreateConfigWithEndpointProvider(S3Request request, String bucket) {
-        AwsRequestOverrideConfiguration requestOverrideConfig =
-            request.overrideConfiguration().orElseGet(() -> AwsRequestOverrideConfiguration.builder().build());
-
-        S3EndpointProvider delegateEndpointProvider = (S3EndpointProvider)
-            requestOverrideConfig.endpointProvider().orElseGet(() -> serviceClientConfiguration().endpointProvider().get());
-
-        return requestOverrideConfig.toBuilder()
-                                    .endpointProvider(BucketEndpointProvider.create(delegateEndpointProvider, bucket))
-                                    .build();
+    private <T extends S3Request, ReturnT> void fetchRegionAndSendRequest(T request,
+                                                                          Function<T, CompletableFuture<ReturnT>> operation,
+                                                                          String bucketName,
+                                                                          CompletableFuture<ReturnT> returnFuture) {
+
+        // // TODO: will fix the casts with separate PR
+        ((S3AsyncClient) delegate()).headBucket(b -> b.bucket(bucketName)).whenComplete((response,
+                                                                                         throwable) -> {
+            if (throwable != null) {
+                if (isS3RedirectException(throwable)) {
+                    bucketToRegionCache.remove(bucketName);
+                    Optional<String> bucketRegion = getBucketRegionFromException((S3Exception) throwable.getCause());
+
+                    if (bucketRegion.isPresent()) {
+                        bucketToRegionCache.put(bucketName, Region.of(bucketRegion.get()));
+                        sendRequestWithRightRegion(request, operation, bucketName, returnFuture, bucketRegion);
+                    } else {
+                        returnFuture.completeExceptionally(throwable);
+                    }
+                } else {
+                    returnFuture.completeExceptionally(throwable);
+                }
+            } else {
+                CompletableFuture<ReturnT> newFuture = operation.apply(request);
+                CompletableFutureUtils.forwardResultTo(newFuture, returnFuture);
+                CompletableFutureUtils.forwardExceptionTo(returnFuture, newFuture);
```

Comment:
What use case does this block of code represent? What is the reason to make an unwrapped API call? We already made one further up the chain.

Reply:
This was just a logical block to handle if headObject Passes .
However , in the given code flow it should never pass since this call is made only when we get a redirect error.